drafts/2018-11-30-zooming-in-on-default-interface-methods.md

```diff
@@ -0,0 +1,454 @@
+
+```csharp
+Order UpdatePostcode(Order order/*👨🏻‍⚖️*/, string newPostcode)
+    => new Order(
+        new Customer(
+            order.Customer.Name,
+            new Address(
+                order.Customer.Address.Line1,
+                order.Customer.Address.Line2,
+                newPostcode
+            )
+        ),
+        order.Products
+    );
+```
+
+F#'s `with`-syntax helps a little, but not a lot. You still have to write each name multiple times.
+
+```fsharp
+let updatePostcode(order : Order, newPostcode : string) : Order = {
+  order with customer = {
+    order.customer with address = {
+      order.customer.address with postcode = newPostcode
+    }
+  }
+}
+```
+
+This pain is caused by immutability. In an imperative setting you can say `order.Customer.Address.Postcode = newPostcode;`, although of course mutable data is less reliable and harder to work with overall. Rather than give up on immutability, functional programmers have invented a remarkable family of composable tools called _optics_ for poking around inside complex datatypes. Optics are a way of describing "paths" through structures: you can compose paths together, and read and write the values at the end of those paths. I'm here today to demonstrate that C#8's upcoming [_default interface methods_](https://github.com/dotnet/csharplang/issues/52) are great at modelling optics.
+
+Let's start with _lenses_, the family member that gave the family its name.
+
+Lenses
+------
+
+A _lens_ is a first-class property for an immutable object. It's an object with a pair of methods, a _getter_ which retrives the value of a property and a _setter_ which updates it. Remember, we're working with immutable data, so the setter returns a new copy of the object.
+
+```csharp
+interface ILens<T, TProp>
+{
+    TProp Get(T obj);
+    T Set(T oldObj, TProp newVal);
+}
+```
+
+For example, here's a lens which focuses on a `Customer`'s `Address`.
+
+```csharp
+class AddressL : ILens<Customer, Address>
+{
+    public Address Get(Customer c) => c.Address;
+
+    public Customer Set(Customer c, Address a) => new Customer(c.Name, a);
+}
+```
+
+**PICTURE**
+
+So a lens picks out a single property inside a given object.
+
+The power of lenses comes from their composability. Given a lens identifying a `T2` inside a `T1` (`ILens<T1, T2>`) and a lens identifying a `T3` inside a `T2` (`ILens<T2, T3>`), you can compose those lenses together to focus all the way from the `T1` to the `T3`.
+
+**PICTURE**
+
+You can traverse any relationship in your data model by composing together a small number of atomic lenses. Composing lenses is so important that I've given it the shortest name I can think of: `_`. (Readers of [an eariler post of mine](2018-03-16-eighty.html) will know of my fondness for `_`.)
```

Comment:
what is an atomic lens?

Reply:
Good question! I'm using "atomic" in the usual English (slash Greek) sense of "indivisible". Do you think I should rephrase?